src/hotspot/share/gc/g1/g1FullCollector.cpp

```diff
@@ -225,15 +228,17 @@ void G1FullCollector::complete_collection() {
   _heap->print_heap_after_full_collection(scope()->heap_transition());
 }
 
-void G1FullCollector::update_attribute_table(HeapRegion* hr) {
-  if (hr->is_free()) {
+void G1FullCollector::update_attribute_table(HeapRegion* hr, bool force_pinned) {
+  if (force_pinned) {
+    _region_attr_table.set_pinned(hr->hrm_index());
     return;
   }
   if (hr->is_closed_archive()) {
     _region_attr_table.set_closed_archive(hr->hrm_index());
   } else if (hr->is_pinned()) {
```

Comment:
Changing the condition to `hr->is_pinned() || force_pinned` is enough for this method, right?

Reply:
not exactly. please check previous discussion at https://github.com/openjdk/jdk/pull/2760#discussion_r602144952.
I will merge the 2 conditions as you suggested.